src/sound/snd_opl_esfm.c

```diff
@@ -0,0 +1,223 @@
+    int8_t  flags;
+    int8_t  pad;
+
+    uint16_t port;
+    uint8_t  status;
+    uint8_t  timer_ctrl;
+    uint16_t timer_count[2];
+    uint16_t timer_cur_count[2];
+
+    // OPL3L
+    int32_t rateratio;
+    int32_t samplecnt;
+    int32_t oldsamples[2];
+    int32_t samples[2];
+
+    int     pos;
+    int32_t buffer[SOUNDBUFLEN * 2];
+} esfm_drv_t;
+
+void
+esfm_drv_generate_resampled(esfm_drv_t *dev, int32_t *bufp)
+{
+    while (dev->samplecnt >= dev->rateratio) {
+        int16_t samples[2] = { 0, 0 };
+        dev->oldsamples[0] = dev->samples[0];
+        dev->oldsamples[1] = dev->samples[1];
+        ESFM_generate(&dev->opl, samples);
+        dev->samples[0] = samples[0];
+        dev->samples[1] = samples[1];
+        dev->samplecnt -= dev->rateratio;
+    }
+
+    bufp[0] = (int32_t) ((dev->oldsamples[0] * (dev->rateratio - dev->samplecnt)
+                          + dev->samples[0] * dev->samplecnt)
+                         / dev->rateratio);
+    bufp[1] = (int32_t) ((dev->oldsamples[1] * (dev->rateratio - dev->samplecnt)
+                          + dev->samples[1] * dev->samplecnt)
+                         / dev->rateratio);
+
+    dev->samplecnt += 1 << RSM_FRAC;
+}
+
+void
+esfm_drv_generate_stream(esfm_drv_t *dev, int32_t *sndptr, uint32_t num)
+{
+    for (uint32_t i = 0; i < num; i++) {
+        esfm_drv_generate_resampled(dev, sndptr);
+        sndptr += 2;
+    }
+}
+
+static void
+esfm_drv_set_do_cycles(void *priv, int8_t do_cycles)
+{
+    esfm_drv_t *dev = (esfm_drv_t *) priv;
+
+    if (do_cycles)
+        dev->flags |= FLAG_CYCLES;
+    else
+        dev->flags &= ~FLAG_CYCLES;
+}
+
+static void *
+esfm_drv_init(const device_t *info)
+{
+    esfm_drv_t *dev = (esfm_drv_t *) calloc(1, sizeof(esfm_drv_t));
+    dev->flags       = FLAG_CYCLES | FLAG_OPL3;
+
+    /* Initialize the ESFMu object. */
+    ESFM_init(&dev->opl);
+    dev->rateratio = (SOUND_FREQ << RSM_FRAC) / 49716;
+
+    return dev;
+}
+
+static void
+esfm_drv_close(void *priv)
+{
+    esfm_drv_t *dev = (esfm_drv_t *) priv;
+    free(dev);
+}
+
+static int32_t *
+esfm_drv_update(void *priv)
+{
+    esfm_drv_t *dev = (esfm_drv_t *) priv;
+
+    if (dev->pos >= sound_pos_global)
+        return dev->buffer;
+
+    esfm_drv_generate_stream(dev,
+                          &dev->buffer[dev->pos * 2],
+                          sound_pos_global - dev->pos);
+
+    for (; dev->pos < sound_pos_global; dev->pos++) {
+        dev->buffer[dev->pos * 2] /= 2;
+        dev->buffer[(dev->pos * 2) + 1] /= 2;
+    }
+
+    return dev->buffer;
+}
+
+
+static void
+esfm_drv_reset_buffer(void *priv)
+{
+    esfm_drv_t *dev = (esfm_drv_t *) priv;
+
+    dev->pos = 0;
+}
+
+static uint8_t
+esfm_drv_read(uint16_t port, void *priv)
+{
+    esfm_drv_t *dev = (esfm_drv_t *) priv;
+
+    if (dev->flags & FLAG_CYCLES)
+        cycles -= ((int) (isa_timing * 8));
+
+    esfm_drv_update(dev);
+
+    uint8_t ret = 0xff;
+
+    ret = ESFM_read_port(&dev->opl, port & 3);
```

Comment:
Is this also reading ESFMu's internal timer implementation? Be sure to test if it's auto-detecting okay in programs, and test for cycle-accuracy if you can, and give me feedback on that, because I've just implemented it and didn't test it yet.

Reply:
I mean as emulated.